lua/nvim-treesitter/textobjects.lua

```diff
@@ -0,0 +1,99 @@
+local api = vim.api
+local ts = vim.treesitter
+
+local configs = require "nvim-treesitter.configs"
+local parsers = require "nvim-treesitter.parsers"
+local queries = require'nvim-treesitter.query'
+local locals = require'nvim-treesitter.locals'
+local ts_utils = require'nvim-treesitter.ts_utils'
+
+local M = {}
+
+function M.select_textobj(query_string)
+  local bufnr = vim.api.nvim_get_current_buf()
+  local ft = api.nvim_buf_get_option(bufnr, "ft")
+  if not ft then return end
+  local lang = parsers.ft_to_lang(ft)
+
+  local row, col = unpack(vim.api.nvim_win_get_cursor(0))
+  row = row - 1
+
+  local matches = {}
+
+  if string.match(query_string, '^@.*') then
+    matches = locals.get_capture_matches(bufnr, query_string, 'textobjects')
+  else
+    local parser = parsers.get_parser(bufnr, lang)
+    local root = parser:parse():root()
+    local start_row, _, end_row, _ = root:range()
+
+    local nested = {}
+    local query = ts.parse_query(lang, query_string)
+    for prepared_match in queries.iter_prepared_matches(query, root, bufnr, start_row, end_row) do
+      table.insert(nested, prepared_match)
+    end
+    for _, m in pairs(nested) do
+      for _, n in pairs(m) do
+        if n.node then
+          table.insert(matches, n.node)
+        end
+      end
+    end
```

Comment:
I think that this can be done in one go :
```suggestion
    for m in queries.iter_prepared_matches(query, root, bufnr, start_row, end_row) do
      for _, n in pairs(m) do
        if n.node then
          table.insert(matches, n.node)
        end
      end
    end
```